### DISCUSSION TITLE
How to reuse and compose shapes within the same file

### DISCUSSION THREAD

**Discussion by sonicsmooth** [terrastruct/d2#2014] [Q&A]
I'd like to be able to reuse a hierarchy of shapes:
For example:

    <def?> mycomponent: {
        the
        box
    }

    mc1: mycomponent
    mc2: mycomponent

should result in 
![image](https://github.com/user-attachments/assets/0d898013-f676-47b1-9da6-7382d0aa8bce)

I can do this by defining mycomponent in another file and then using ...@filename, but I'd like to keep the component definition in the main file.

It seems like classes can be used to compose properties and styles, but not shapes.
eg [https://github.com/terrastruct/d2/discussions/1087](url)

This doesn't work:

    classes: {
        mycomponent: {
            some
            thing
        }
    }

    x.class: mycomponent
    y.class: mycomponent

It seems like vars can't contain shapes.

Perhaps I'm missing something obvious with these.

thanks

**Comment by alixander** [comment]
vars can do this. unless i'm misunderstanding, this is the same issue as https://github.com/terrastruct/d2/issues/1891#issuecomment-2237737637

**Reply by sonicsmooth** [reply]
Thanks, that's close.
How do I use vars to collect a number of shapes+arrows into one thing, that can then get instantiated many times?

Like below, where I don't have to type dog1, dog3, etc., multiple times.

![image](https://github.com/user-attachments/assets/9cbf7b92-9117-40f2-9abd-d452418198a0)

    my-house: {
      ...${my-house}
      dog1 -- dog3: "Couple"
      dog1: ${dog1}
      dog3: ${dog3}
}

**Reply by alixander** [reply]
```d2
vars: {
  house: {
    cat -> dog
  }
}

mansion: {
  ...${house}
}

shed: {
  ...${house}
}
```

![Screenshot 2024-07-24 at 8 28 52 AM](https://github.com/user-attachments/assets/9b9ab36b-cf65-46f8-b0b4-db8482f329ba)

**Reply by sonicsmooth** [reply]
Thanks!  This is the obvious thing I was missing :)
Can a var refer to another var in the same scope?

    vars: {
        Apple: {
            shape:circle
            label:Granny Smith
        }
        Cherry: {
            shape:circle
            label:Rainier Cherry
        }
        SummerFruit: {
            xx: ${Apple}   # Could not resolve
            cc: ${Cherry}  # Could not resolve
            xx -> cc
        }
    }

    x: ${Apple}   # ok
    c: ${Cherry}  # ok
    x -> c
    sf: ${SummerFruit} # doesn't get this far

**Reply by alixander** [reply]
It should, I made an issue: https://github.com/terrastruct/d2/issues/2017
